assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl

```diff
@@ -1,33 +1,30 @@
 ---
 systemd:
   units:
-    - name: etcd-member.service
+    - name: etcd.service
       enable: true
-      dropins:
-        - name: 40-etcd-cluster.conf
-          contents: |
-            [Service]
-            Environment="ETCD_IMAGE_TAG=v3.4.13"
-            Environment="ETCD_IMAGE_URL=docker://quay.io/coreos/etcd"
-            Environment="RKT_RUN_ARGS=--insecure-options=image"
-            Environment="ETCD_NAME=${etcd_name}"
-            Environment="ETCD_ADVERTISE_CLIENT_URLS=https://${etcd_domain}:2379"
-            Environment="ETCD_INITIAL_ADVERTISE_PEER_URLS=https://${etcd_domain}:2380"
-            Environment="ETCD_LISTEN_CLIENT_URLS=https://0.0.0.0:2379"
-            Environment="ETCD_LISTEN_PEER_URLS=https://0.0.0.0:2380"
-            Environment="ETCD_LISTEN_METRICS_URLS=http://0.0.0.0:2381"
-            Environment="ETCD_INITIAL_CLUSTER=${etcd_initial_cluster}"
-            Environment="ETCD_STRICT_RECONFIG_CHECK=true"
-            Environment="ETCD_SSL_DIR=/etc/ssl/etcd"
-            Environment="ETCD_TRUSTED_CA_FILE=/etc/ssl/certs/etcd/server-ca.crt"
-            Environment="ETCD_CERT_FILE=/etc/ssl/certs/etcd/server.crt"
-            Environment="ETCD_KEY_FILE=/etc/ssl/certs/etcd/server.key"
-            Environment="ETCD_CLIENT_CERT_AUTH=true"
-            Environment="ETCD_PEER_TRUSTED_CA_FILE=/etc/ssl/certs/etcd/peer-ca.crt"
-            Environment="ETCD_PEER_CERT_FILE=/etc/ssl/certs/etcd/peer.crt"
-            Environment="ETCD_PEER_KEY_FILE=/etc/ssl/certs/etcd/peer.key"
-            Environment="ETCD_PEER_CLIENT_CERT_AUTH=true"
-            ExecStopPost=-/opt/etcd-rejoin
+      contents: |
+        [Unit]
+        Description=etcd (System Application Container)
+        Documentation=https://github.com/etcd-io/etcd
+        Wants=docker.service
+        After=docker.service
+        [Service]
+        Type=simple
+        Restart=always
+        RestartSec=10s
+        TimeoutStartSec=0
+        LimitNOFILE=40000
+        EnvironmentFile=/etc/kubernetes/etcd.env
+        ExecStart=sh -c "docker run --network=host \
+          -u $(id -u \"$${ETCD_USER}\"):$(id -u \"$${ETCD_USER}\") \
```

Comment:
Can't we remove the `sh -c`?

I guess it is used  for the `id -u` part. But docker takes a string for `-u` param. Do we really need to run that? Even if we do, can't we add a variable for that in the env file?

Reply:
Yeah, I think it should be possible to remove it. Good point @rata

Reply:
tl;dr; string user names don't work on flatcar.

---

On Flatcar when I do this I get the user id:

```console
# id -u etcd
232
```

But there is no entry for that user in `/etc/passwd`:

```console
# cat /etc/passwd
root:x:0:0:root:/root:/bin/bash
core:x:500:500:Flatcar Admin:/home/core:/bin/bash
systemd-timesync:x:997:997:systemd Time Synchronization:/:/sbin/nologin
systemd-coredump:x:996:996:systemd Core Dumper:/:/sbin/nologin
```

So docker fails:

```console
# docker run -u etcd fedora bash
/run/torcx/bin/docker: Error response from daemon: linux spec user: unable to find user etcd: no matching entries in passwd file.
ERRO[0000] error waiting for container: context canceled
```

There is a workaround you can use which is `echo 'etcd:x:232:232::/dev/null:/sbin/nologin' | sudo tee -a /etc/passwd` but this does not work as well.

So there is another passwd file at play here: https://github.com/flatcar-linux/baselayout/blob/flatcar-master/baselayout/passwd, this is where we get the user id of etcd from.

Reply:
>There is a workaround you can use which is echo 'etcd:x:232:232::/dev/null:/sbin/nologin' | sudo tee -a /etc/passwd but this does not work as well.

Can't we create the user using ignition too?

Reply:
Added following line, but there was no effect:

```diff
diff --git a/assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl b/assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controll
er.yaml.tmpl
index fd50be82a..e911708ee 100644
--- a/assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
+++ b/assets/terraform-modules/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl
@@ -252,3 +252,4 @@ passwd:
   users:
     - name: core
       ssh_authorized_keys: ${ssh_keys}
+    - name: etcd
```

on the controller host:

```console
# cat /etc/passwd
root:x:0:0:root:/root:/bin/bash
core:x:500:500:Flatcar Admin:/home/core:/bin/bash
systemd-timesync:x:997:997:systemd Time Synchronization:/:/sbin/nologin
systemd-coredump:x:996:996:systemd Core Dumper:/:/sbin/nologin
```

Also if I manually try to add the user it does not work, because the user already exists:

```console
# useradd etcd
useradd: user 'etcd' already exists
```

The user is already there in:

```
# cat /usr/share/baselayout/passwd | grep etcd
etcd:x:232:232::/dev/null:/sbin/nologin
```

But docker does not identify this, hence the workaround.

I had a chat with Thilo and this is what he had to add:

> Baselayout provides `/etc/passwd` for the `initrd`, and includes the `etcd` user. The `/etc/passwd` is the runtime configuration, which is generated from the systemd build recipe.
> 
> - baselayout (w/ `etcd` user: https://github.com/flatcar-linux/baselayout/blob/flatcar-master/baselayout/passwd).
> - systemd build recipe in `coreos-overlay`: https://github.com/flatcar-linux/coreos-overlay/blob/main/sys-apps/systemd/systemd-9999.ebuild#L534-L542
> 
> For a quick patch, please try and work around this by issuing
> 
> ```bash
> echo 'etcd:x:232:232::/dev/null:/sbin/nologin' | sudo tee -a /etc/passwd
> ```
> 
> In the long term we should really only have one authoritative source for passwd.

Reply:
That's so odd :smile: Maybe we should do something like `grep etcd /usr/share/baselayout/passwd | sudo tee -a /etc/passwd` then?

Reply:
Like I said before in https://github.com/kinvolk/lokomotive/pull/946#discussion_r489395018 that that does not work.

Reply:
I guess it's okay then. Thanks for explanation @surajssd. @rata ?

Reply:
Accidentally clicked Resolve. So this means that the etcd container creates files owned by the `etcd` Flatcar user?

Reply:
As far as I understand, yes.

Reply:
The UID numbers are same so yeah.

Reply:
IMHO, I'd either dig deeper to understand why we can't properly use the etcd user on Flatcar (and create a bug report if needed) or create an issue to investigate later (I lean towards the first, though :))

If we want to investigate it later, I think it is cleaner to create a variable in the env-file with the UID and use that variable here, instead of running a shell to run `id -u`.

Reply:
>If we want to investigate it later, I think it is cleaner to create a variable in the env-file with the UID and use that variable here, instead of running a shell to run id -u.

That's a lot of indirection then to have separate unit to write env file etc. IMO the current way is simpler.

Reply:
Oh, right, if we need a new unit for this (I thought we had one to create those things already) then this way is not so bad :)

Reply:
so can we resolve this one?

Reply:
I think @rata suggested to either investigate deeper why we can't properly use etcd user on Flatcar OR create an issue for it to investigate later (though honestly, this seems like an issue we will never go back to, but I don't mind having it created).

And as creating new unit to create env file seems complicated, I have a feeling that currently approach with `sh` is okay'ish.

Reply:
Docker takes the user ID from the image itself. If the image specifies an etcd user, this has probably not the same user ID as the one Flatcar uses, therefore, the current way makes sense as long as you want to depend on the basic Flatcar etcd setup.